### PR TITLE
update model server support

### PR DESCRIPTION
**Pull request by sunya-ch**
This PR updates model server support aiming for release v0.6 as mentioned in https://github.com/sustainable-computing-io/kepler-operator/issues/232. 

API doc: 
https://github.com/sustainable-computing-io/kepler-operator/blob/c15c77621958cc79d1921d9af378915158abc4ca/docs/api.md

The PR contains changes on :
- updating KeplerSpec
- updating modelserver component
- adding estimator component
- modifying exporter to have sidecar if configure
- adding modelServerReconcilers

Note that, 
The holder for setting filters and model name is here on kepler: 
https://github.com/sustainable-computing-io/kepler/blob/73cb11fb963f425013cf7f03f214c8f8b85c7853/pkg/config/config.go#L390.
However, it is not determined how to use it. So, it is not supported yet from end to end.

Example configmap change from full deployment on OpenShift on IBM Cloud 
(kepler CR: `config/samples/kepler_full_deploy.yaml`)

```yaml
> oc get configmap -n openshift-kepler-operator model-server-cm -oyaml
...
  MODEL_CONFIG: |
    NODE_COMPONENTS_ESTIMATOR=true
    NODE_COMPONENTS_INIT_URL=https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-db/main/models/Linux-4.15.0-213-generic-x86_64_v0.6/rapl/AbsPower/KubeletOnly/GradientBoostingRegressorTrainer_1.zip
  MODEL_SERVER_ENABLE: "true"
  MODEL_SERVER_URL: http://model-server-svc.openshift-kepler-operator.svc.cluster.local:8100
```

Resources:

```bash
> oc get -n openshift-kepler-operator all
NAME                                       READY   STATUS    RESTARTS   AGE
pod/kepler-exporter-ds-48lff               2/2     Running   0          4m10s
pod/kepler-exporter-ds-4nj8g               2/2     Running   0          4m10s
pod/kepler-exporter-ds-5lj62               2/2     Running   0          4m10s
pod/kepler-exporter-ds-9rqlv               2/2     Running   0          4m10s
pod/kepler-exporter-ds-knnsh               2/2     Running   0          4m10s
pod/kepler-exporter-ds-kskmq               2/2     Running   0          4m11s
pod/kepler-exporter-ds-szl8w               2/2     Running   0          2m49s
pod/model-server-deploy-85fd7b8c6d-wkkrg   1/1     Running   0          4m10s

NAME                          TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
service/kepler-exporter-svc   ClusterIP   None         <none>        9103/TCP   4m18s
service/model-server-svc      ClusterIP   None         <none>        8100/TCP   4m17s

NAME                                DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/kepler-exporter-ds   7         7         7       7            7           kubernetes.io/os=linux   4m19s

NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/model-server-deploy   1/1     1            1           4m18s

NAME                                             DESIRED   CURRENT   READY   AGE
replicaset.apps/model-server-deploy-85fd7b8c6d   1         1         1       4m18s
```

## exporter log
```
> oc logs -n openshift-kepler-operator kepler-exporter-ds-szl8w estimator
I0914 08:42:05.585412 3734085 node_component_energy.go:54] Using the EstimatorSidecar/AbsPower Power Model to estimate Node Component Power
```

## estimator log
```
> oc logs -n openshift-kepler-operator kepler-exporter-ds-szl8w estimator
set NODE_COMPONENTS_ESTIMATOR to true.
set NODE_COMPONENTS_INIT_URL to https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-db/main/models/Linux-4.15.0-213-generic-x86_64_v0.6/rapl/AbsPower/KubeletOnly/GradientBoostingRegressorTrainer_1.zip.
clean socket
load model from model server:  /mnt/download/rapl/AbsPower
```

Signed-off-by: Sunyanan Choochotkaew <sunyanan.choochotkaew1@ibm.com>

### CODE REVIEW

**Comment by sunya-ch** [comment]
Please feel free to review on the design first. 
I'm working on fixing the syntax issue. Will amend the commit once confirm all deployment choices work at least on my cluster.

**Comment by sunya-ch** [comment]
Now fix critical issue on deployment. 

However, please allow me to have some issue left (need help from other to fix with other PR):
- Kepler model server container created after exporter/estimator --> cannot load model at the beginning, need to restart exporter pod after model server pod is ready manually.
- Integration test (now manually deploy on owned cluster)

**Comment by sthaha** [comment]
@sunya-ch Thanks a lot for adding the feature 🤗 
Please allow us some time to go through the feature implementation. My first focus will be on the `spec.modelServer` to ensure we have only the minimal set of api exposed.

File: cmd/manager/main.go [hunk 1]
```diff
@@ -86,6 +88,12 @@ func main() {
 	keplerImage := env("KEPLER_IMAGE", exporter.StableImage)
 	flag.StringVar(&exporter.Config.Image, "kepler.image", keplerImage, "kepler image")
 
+	keplerEstimatorImage := env("KEPLER_ESTIMATOR_IMAGE", estimator.StableImage)
+	flag.StringVar(&estimator.Config.Image, "kepler-estimator.image", keplerEstimatorImage, "kepler estimator image")
+
+	keplerModelServerImage := env("KEPLER_MODEL_SERVER_IMAGE", modelserver.StableImage)
+	flag.StringVar(&modelserver.Config.Image, "kepler-model-server.image", keplerModelServerImage, "kepler model server image")
```
Comment (sthaha) [inline cmd/manager/main.go:95]: (suggestion): we have an opportunity to refactor this pattern of getting the env from the flag.
Lets use `estimator.image` and `model-server.image` in the flag please. 

I have a question about the name `model-server`. Do you know why we call it that? what does it really serve?

File: pkg/api/v1alpha1/kepler_types.go [hunk 2]
```diff
@@ -114,7 +129,12 @@ type ExporterSpec struct {
 
 // KeplerSpec defines the desired state of Kepler
 type KeplerSpec struct {
-	Exporter ExporterSpec `json:"exporter,omitempty"`
+	Exporter                     ExporterSpec     `json:"exporter,omitempty"`
+	ModelServer                  *ModelServerSpec `json:"modelServer,omitempty"`
+	NodeTotalEstimator           *EstimatorSpec   `json:"nodeTotalEstimator,omitempty"`
+	NodeComponentsEstimator      *EstimatorSpec   `json:"nodeComponentsEstimator,omitempty"`
+	ContainerTotalEstimator      *EstimatorSpec   `json:"containerTotalEstimator,omitempty"`
+	ContainerComponentsEstimator *EstimatorSpec   `json:"containerComponentsEstimator,omitempty"`
```
Comment (sthaha) [inline pkg/api/v1alpha1/kepler_types.go:137]: How about we keep all `estimator` related spec under

```yaml
spec:
   estimator:
      node:
         total
         components

      container
          total:
          components:
```

File: pkg/components/estimator/estimator.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+// Config that will be set from outside
+var (
+	Config = struct {
+		Image string
+	}{
+		Image: StableImage,
+	}
+
+	bashCommand = []string{"/bin/sh", "-c"}
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:47]: Lets keep all private (not exported) in  the `var` section below.

File: pkg/components/estimator/estimator.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+var (
+	// common labels for all resources of modelserver
+	labels = components.CommonLabels.Merge(k8s.StringMap{
+		"app.kubernetes.io/component":  "model-server",
+		"sustainable-computing.io/app": "model-server",
+	})
+)
+
+// IsEstimatorSidecarEnabled returns true if any of estimator config has sidecar enabled
+func IsEstimatorSidecarEnabled(k *v1alpha1.Kepler) bool {
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:59]: How about ... 

```suggestion
func NeedsEstimatorSidecar(es *v1alpha1.EstimatorSpec) bool {
```

File: pkg/components/estimator/estimator.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+// NewContainer returns sidecar additional volumes
+func NewVolumes() []corev1.Volume {
+	return []corev1.Volume{
+		k8s.VolumeFromEmptyDir("mnt"),
+		k8s.VolumeFromEmptyDir("tmp"),
+	}
+}
+
+func GetModifiedVolumeMount(volumeMounts []corev1.VolumeMount) []corev1.VolumeMount {
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:107]:
```suggestion
func AddTmpMount(volumeMounts []corev1.VolumeMount) []corev1.VolumeMount {
```

File: pkg/components/estimator/estimator.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		Name:      "tmp",
+		MountPath: "/tmp",
+	})
+}
+
+func GetModifiedCommandAndArgs(inCmd []string) (outCmd, outArgs []string) {
+	outCmd = bashCommand
+	outArgs = []string{fmt.Sprintf(waitForSocketCommand, strings.Join(inCmd, " "))}
+	return
+}
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:118]:
```suggestion
func estimatorCmdAndArgs(inCmd []string) ([]string, []string) {
	return (
	  bashCommand
	  []string{fmt.Sprintf(waitForSocketCommand, strings.Join(inCmd, " "))}
	)
}
```

File: pkg/components/estimator/estimator.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+func getEstimatorConfig(prefix string, spec v1alpha1.EstimatorSpec) string {
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:120]:
```suggestion
func estimatorConfig(prefix string, spec v1alpha1.EstimatorSpec) string {
```
Getters - https://go.dev/doc/effective_go#Getters  `GetFoobar` -> `Foobar`

File: pkg/components/estimator/estimator.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	config := fmt.Sprintf("%s_ESTIMATOR=%v\n", prefix, spec.SidecarEnabled)
+	if spec.InitUrl != "" {
+		config += fmt.Sprintf("%s_INIT_URL=%s\n", prefix, spec.InitUrl)
+	}
+	return config
+}
+
+func GetModelConfig(k *v1alpha1.Kepler) string {
```
Comment (sthaha) [inline pkg/components/estimator/estimator.go:128]:
```suggestion
func ModelConfig(es v1alpha1.EstimatorSpec) string {
```
2 suggestions 
1. Getters - https://go.dev/doc/effective_go#Getters  `GetFoobar` -> `Foobar` 
2. Pass the minimal required information to the function. In this case, we

File: pkg/components/estimator/estimator_test.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/sustainable.computing.io/kepler-operator/pkg/api/v1alpha1"
+	"github.com/sustainable.computing.io/kepler-operator/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func getEnaledSidecarSpec() *v1alpha1.EstimatorSpec {
```
Comment (sthaha) [inline pkg/components/estimator/estimator_test.go:12]:
```suggestion
func sidecarEnabledSpec() *v1alpha1.EstimatorSpec {
```

File: pkg/components/estimator/estimator_test.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	return &v1alpha1.EstimatorSpec{
+		InitUrl:        "fake-url.zip",
+		SidecarEnabled: true,
+	}
+}
+
+func getDisabledSidecarSpec() *v1alpha1.EstimatorSpec {
```
Comment (sthaha) [inline pkg/components/estimator/estimator_test.go:19]:
```suggestion
func sidecarDisabledSpec() *v1alpha1.EstimatorSpec {
```

File: pkg/components/estimator/estimator_test.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+			scenario:  "default case",
+		},
+		{
+			spec: &v1alpha1.KeplerSpec{
+				NodeTotalEstimator:           getDisabledSidecarSpec(),
+				NodeComponentsEstimator:      getDisabledSidecarSpec(),
+				ContainerTotalEstimator:      getDisabledSidecarSpec(),
+				ContainerComponentsEstimator: getDisabledSidecarSpec(),
+			},
+			configStr: nodeTotalDisaled + nodeComponentsDisabled + containerTotalDisaled + containerComponentsDisabled,
```
Comment (sthaha) [inline pkg/components/estimator/estimator_test.go:57]: typo in `Disa**b**led`

File: pkg/components/estimator/estimator_test.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+func getDisabledSidecarSpec() *v1alpha1.EstimatorSpec {
+	return &v1alpha1.EstimatorSpec{
+		InitUrl:        "fake-url.json",
+		SidecarEnabled: false,
+	}
+}
+
+func TestModelConfig(t *testing.T) {
+
+	nodeTotalEnaled := "NODE_TOTAL_ESTIMATOR=true\nNODE_TOTAL_INIT_URL=fake-url.zip\n"
```
Comment (sthaha) [inline pkg/components/estimator/estimator_test.go:28]:
```suggestion
	nodeTotalEnabled := "NODE_TOTAL_ESTIMATOR=true\nNODE_TOTAL_INIT_URL=fake-url.zip\n"
```
typo `Ena*b*led`

File: pkg/components/exporter/exporter.go [hunk 13]
```diff
@@ -106,6 +108,74 @@ func NewDaemonSet(detail components.Detail, k *v1alpha1.Kepler) *appsv1.DaemonSe
 
 	bindAddress := "0.0.0.0:" + strconv.Itoa(int(deployment.Port))
 
+	exporterContainer := &corev1.Container{
```
Comment (sthaha) [inline pkg/components/exporter/exporter.go:111]:
```suggestion
	exporterContainer := newExporterContainer(k)
```
Lets move all the creation to a `newExporterContainer()` which can be unit tested as well.

File: pkg/components/exporter/exporter.go [hunk 14]
```diff
@@ -106,6 +108,74 @@ func NewDaemonSet(detail components.Detail, k *v1alpha1.Kepler) *appsv1.DaemonSe
...
+	var volumes = []corev1.Volume{
+		k8s.VolumeFromHost("lib-modules", "/lib/modules"),
+		k8s.VolumeFromHost("tracing", "/sys"),
+		k8s.VolumeFromHost("proc", "/proc"),
+		k8s.VolumeFromHost("kernel-src", "/usr/src/kernels"),
+		k8s.VolumeFromHost("kernel-debug", "/sys/kernel/debug"),
+		k8s.VolumeFromConfigMap("cfm", ConfigmapName),
+	}
+
+	if estimator.IsEstimatorSidecarEnabled(k) {
```
Comment (sthaha) [inline pkg/components/exporter/exporter.go:168]:
```suggestion
         containers := []corev1.Container{*exporterContainer}
         
	if estimator.NeedsSidecar(k) {
	    containers = addEstimatorSidecar(exporterContainer)
	}
```

File: pkg/api/v1alpha1/kepler_types.go [hunk 15]
```diff
@@ -58,16 +64,13 @@ type ModelServerTrainerSpec struct {
 
 	// +kubebuilder:default=true
 	PromSSLDisable bool `json:"promSSLDisable,omitempty"`
```
Comment (sthaha) [inline pkg/api/v1alpha1/kepler_types.go:66]: How about we move all Prometheus Config to separate struct?

File: pkg/components/modelserver/modelserver.go [hunk 16]
```diff
@@ -168,14 +176,29 @@ func NewService(k *v1alpha1.Kepler) *corev1.Service {
...
+		"MODEL_SERVER_URL": defaultIfEmpty(ms.URL, GetDefaultModelServer(*k.Spec.ModelServer)),
 	}
+	msConfig = msConfig.AddIfNotEmpty("MODEL_SERVER_REQ_PATH", ms.RequestPath)
+	msConfig = msConfig.AddIfNotEmpty("MODEL_SERVER_MODEL_LIST_PATH", ms.ListPath)
+
+	return msConfig.ToMap()
 }
 
-func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) *corev1.ConfigMap {
+func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) (*corev1.ConfigMap, error) {
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:201]:
```suggestion
func NewConfigMap(d components.Detail, ms v1alpha1.ModelServerSpec) (*corev1.ConfigMap) {
```

Returing error only helps to mask logical error (made by the developer) which is that this function should not  be called if modelserver is nil. By changing the arg to accept an object instead of pointer, we don't have to return error (keeping the signature of all New<X> the same).

File: pkg/components/modelserver/modelserver.go [hunk 17]
```diff
@@ -74,21 +73,11 @@ var (
...
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: corev1.SchemeGroupVersion.String(),
-			Kind:       "ServiceAccount",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      ServiceAccountName,
-			Namespace: components.Namespace,
-			Labels:    labels,
-		},
+func NewDeployment(k *v1alpha1.Kepler) (*appsv1.Deployment, error) {
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:76]:
```suggestion
func NewDeployment(k *v1alpha1.Kepler) (*appsv1.Deployment) {
```

File: pkg/components/modelserver/modelserver.go [hunk 18]
```diff
@@ -74,21 +73,11 @@ var (
...
+	ms := k.Spec.ModelServer
+	if ms == nil {
+		return nil, utils.ErrNoConfig
 	}
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:80]: I think we shouldn't mask logical error which is that these functions should not be called when `modelserver` is nil.

**Review by sthaha — CHANGES_REQUESTED** [review]

**Comment by sthaha** [comment]
@sunya-ch , Thanks a lot for adding this feature 🙇 . 
You can ignore most of the comments in the review, lets focus on getting the `spec.modelserver` and `spec.estimator` parts to the **minimal** required configuration. We should be able to make assumptions about the model server that is deployed, and thus may not need all the configurations currently in place.

File: cmd/manager/main.go (hunk 1 above)
Comment (sunya-ch) [reply cmd/manager/main.go:95]: model-server serves the model selection based on available features. When online-trainer is validated, it will serve the fine-tuned model online to the exporter.

File: pkg/components/exporter/exporter.go (hunk 14 above)
Comment (sunya-ch) [reply pkg/components/exporter/exporter.go:168]: As it also needs changes on volume, I modify your suggestion as below. 

```go
containers, volumes = addEstimatorSidecar(exporterContainer, volumes)
```

File: pkg/components/estimator/estimator.go (hunk 6 above)
Comment (sunya-ch) [reply pkg/components/estimator/estimator.go:118]: Previously, it is used by exporter module. Now, put it (together with tmp volume mount) inside  newly-defined `AddEstimatorDependency` function.

**Comment by sunya-ch** [comment]
@sthaha Thank you so much for the review.
I made most changes according to your review. I put comment below the review that is modified slightly from your suggestion.

File: pkg/components/modelserver/modelserver.go [hunk 19]
```diff
@@ -154,10 +144,17 @@ func NewDeployment(k *v1alpha1.Kepler) *appsv1.Deployment {
...
 }
 
-func NewService(k *v1alpha1.Kepler) *corev1.Service {
-	exporterPort := int32(k.Spec.Exporter.Port)
+func NewService(ms *v1alpha1.ModelServerSpec) *corev1.Service {
+	port := defaultPort
+	if ms.Port > 0 {
+		port = ms.Port
+	}
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:151]: this shouldn't be needed since the api server should reject anything less than 1

File: pkg/components/modelserver/modelserver.go [hunk 20]
```diff
@@ -216,35 +219,47 @@ func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) *corev1.ConfigMap {
...
 }
 
 func settingsForTrainer(trainer *v1alpha1.ModelServerTrainerSpec) k8s.StringMap {
-	if trainer == nil {
-		return nil
-	}
+	msConfig := k8s.StringMap{}
+
+	prom := trainer.Prom
+	if prom != nil {
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:230]: (nit): prefer early returns for better readability 

```suggestion
	if prom == nil {
	   return msConfig
	}
```

File: pkg/components/modelserver/modelserver.go [hunk 21]
```diff
@@ -216,35 +219,47 @@ func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) *corev1.ConfigMap {
...
+		promHeaders := strings.Builder{}
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:231]: nice!

File: pkg/components/modelserver/modelserver.go [hunk 22]
```diff
@@ -216,35 +219,47 @@ func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) *corev1.ConfigMap {
...
+		interval := defaultPromInterval
+		step := defaultPromStep
+		// iterate through available headers and append to promHeaders (this string will be converted to dict in the image)
+		for _, h := range prom.Headers {
+			promHeaders.WriteString(h.Key)
+			promHeaders.WriteString(":")
+			promHeaders.WriteString(h.Value)
+			promHeaders.WriteString(",")
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:239]: is a trailing comma ok?

File: pkg/components/modelserver/modelserver.go (hunk 19 above)
Comment (sunya-ch) [reply pkg/components/modelserver/modelserver.go:151]: I don't see webhook that check Kepler CR?
I need this at least for the unit test because it is not filled by kbuilder as in csv file. 
Even though, in deployment, if user does not define it will be 8100, I just have this in case to make sure we set the default port if it is not properly defined.

File: pkg/utils/k8s/k8s.go [hunk 23]
```diff
@@ -61,6 +61,13 @@ func (l StringMap) ToMap() map[string]string {
 	return l
 }
 
+func (l StringMap) AddIfNotEmpty(k, v string) map[string]string {
```
Comment (sthaha) [inline pkg/utils/k8s/k8s.go:64]: Should this return StringMap instead? 
```suggestion
func (l StringMap) AddIfNotEmpty(k, v string) StringMap {
```

File: pkg/components/modelserver/modelserver.go (hunk 20 above)
Comment (sunya-ch) [reply pkg/components/modelserver/modelserver.go:230]: I intentionally not return it at the beginning for the future case when we have more setting to the trainer not only prom.

File: pkg/controllers/kepler.go [hunk 24]
```diff
@@ -425,9 +434,32 @@ func exporterReconcilers(k *v1alpha1.Kepler, cluster k8s.Cluster) []reconciler.R
...
+		cm,
+		deploy,
+		svc,
+		pvc,
+	)
+
+	if cluster == k8s.OpenShift {
+		updater := newUpdaterForKepler(k)
+		rs = append(rs, updater(exporter.NewSCC(components.Full, k)))
+	}
```
Comment (sthaha) [inline pkg/controllers/kepler.go:459]:
```suggestion
```

File: pkg/components/modelserver/modelserver.go [hunk 25]
```diff
@@ -168,14 +165,24 @@ func NewService(k *v1alpha1.Kepler) *corev1.Service {
...
-				Port:       exporterPort,
+				Port:       int32(port),
 				TargetPort: intstr.FromString("http"),
 			}},
 		},
 	}
 }
 
-func NewConfigMap(d components.Detail, k *v1alpha1.Kepler) *corev1.ConfigMap {
+func GetModelServerConfigForClient(ms *v1alpha1.ModelServerSpec) map[string]string {
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:175]:
```suggestion
func ModelServerConfigForClient(ms v1alpha1.ModelServerSpec) k8s.StringMap {
```
Keeping it k8s.StringMap has the advantage that we can merge 2 or more maps.

File: pkg/components/modelserver/modelserver.go [hunk 26]
```diff
@@ -274,3 +289,11 @@ func defaultIfEmpty(given, defaultVal string) string {
 	}
 	return defaultVal
 }
+
+func GetDefaultModelServer(ms v1alpha1.ModelServerSpec) string {
```
Comment (sthaha) [inline pkg/components/modelserver/modelserver.go:293]:
```suggestion
func serverUrl(ms v1alpha1.ModelServerSpec) string {
```

File: pkg/components/modelserver/modelserver.go (hunk 19 above)
Comment (sthaha) [reply pkg/components/modelserver/modelserver.go:151]: We don't need webhook if we add the right markers 

```
// +kubebuilder:validation:Minimum
specifies the minimum numeric value that this field can have. Negative numbers are supported.
```
https://book.kubebuilder.io/reference/markers/crd-validation.html 
see: exporter.deployment.port

This also reminds me that we should also separate out `deployment` related settings to a `DeploymentSpec`

File: pkg/components/modelserver/modelserver.go (hunk 20 above)
Comment (sthaha) [reply pkg/components/modelserver/modelserver.go:230]: Please feel free to ignore `nits`. 
imho, it is better to write the code for the current scenario and readjust if/when there is a change in future 👼 (YAGNI) :)

File: pkg/utils/k8s/k8s.go [hunk 27]
```diff
@@ -61,6 +61,13 @@ func (l StringMap) ToMap() map[string]string {
...
+	if v != "" {
```
Comment (sthaha) [inline pkg/utils/k8s/k8s.go:65]: should we check the key `k` as well?

**Comment by sunya-ch** [comment]
Made an update to the review that marked the icon. 

Here are example deployments.

## minimum deployment
```yaml
spec:
  exporter:
    deployment:
      port: 9103
```

```
oc get -n openshift-kepler-operator all
NAME                           READY   STATUS    RESTARTS   AGE
pod/kepler-exporter-ds-d4ctn   1/1     Running   0          11s
pod/kepler-exporter-ds-fd5xt   1/1     Running   0          11s
pod/kepler-exporter-ds-fzjk7   1/1     Running   0          11s
pod/kepler-exporter-ds-n46xf   1/1     Running   0          11s
pod/kepler-exporter-ds-nthsc   1/1     Running   0          11s
pod/kepler-exporter-ds-qm7p4   1/1     Running   0          11s
pod/kepler-exporter-ds-s5t48   1/1     Running   0          11s

NAME                          TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
service/kepler-exporter-svc   ClusterIP   None         <none>        9103/TCP   11s

NAME                                DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/kepler-exporter-ds   7         7         7       7            7           kubernetes.io/os=linux   11s
```

## with estimator only
```yaml
spec:
  exporter:
    deployment:
      port: 9103
  estimator:
    node:
      components:
        sidecar: true
        initUrl: https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-db/main/models/Linux-4.15.0-213-generic-x86_64_v0.6/rapl/AbsPower/KubeletOnly/GradientBoostingRegressorTrainer_1.zip
```
```
NAME                           READY   STATUS    RESTARTS   AGE
pod/kepler-exporter-ds-5g5kk   2/2     Running   0          16s
pod/kepler-exporter-ds-7tg9j   2/2     Running   0          16s
pod/kepler-exporter-ds-fh4f2   2/2     Running   0          16s
pod/kepler-exporter-ds-fqdnf   2/2     Running   0          16s
pod/kepler-exporter-ds-lgfwx   2/2     Running   0          16s
pod/kepler-exporter-ds-nthhd   2/2     Running   0          16s
pod/kepler-exporter-ds-pgrl6   2/2     Running   0          16s

NAME                          TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
service/kepler-exporter-svc   ClusterIP   None         <none>        9103/TCP   16s

NAME                                DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/kepler-exporter-ds   7         7         7       7            7           kubernetes.io/os=linux   17s
```

## full deployment
```yaml
spec:
  exporter:
    deployment:
      port: 9103
  estimator:
    node:
      components:
        sidecar: true
        initUrl: https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-db/main/models/Linux-4.15.0-213-generic-x86_64_v0.6/rapl/AbsPower/KubeletOnly/GradientBoostingRegressorTrainer_1.zip
  modelServer:
    enabled: true
```

```
oc get all -n openshift-kepler-operator
NAME                                       READY   STATUS    RESTARTS   AGE
pod/kepler-exporter-ds-4bsnt               2/2     Running   0          4m48s
pod/kepler-exporter-ds-679tv               2/2     Running   0          4m48s
pod/kepler-exporter-ds-6cmkf               2/2     Running   0          4m48s
pod/kepler-exporter-ds-9ltv4               2/2     Running   0          4m49s
pod/kepler-exporter-ds-c6wnl               2/2     Running   0          4m48s
pod/kepler-exporter-ds-f2l9z               2/2     Running   0          4m49s
pod/kepler-exporter-ds-z5wkg               2/2     Running   0          2m55s
pod/model-server-deploy-85fd7b8c6d-9dwz9   1/1     Running   0          4m49s

NAME                          TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
service/kepler-exporter-svc   ClusterIP   None         <none>        9103/TCP   4m50s
service/model-server-svc      ClusterIP   None         <none>        8100/TCP   4m50s

NAME                                DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/kepler-exporter-ds   7         7         7       7            7           kubernetes.io/os=linux   4m50s

NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/model-server-deploy   1/1     1            1           4m50s

NAME                                             DESIRED   CURRENT   READY   AGE
replicaset.apps/model-server-deploy-85fd7b8c6d   1         1         1       4m50s
```

File: pkg/components/modelserver/modelserver.go (hunk 22 above)
Comment (sunya-ch) [reply pkg/components/modelserver/modelserver.go]: I'm not sure who wrote this part. It will be used here: https://github.com/sustainable-computing-io/kepler-model-server/blob/4af262f2bb2ae2420eaddf64cdc8480eb05ff3e6/src/train/prom/prom_query.py#L26. 
Could I left someone work on that later if not match?

Comment (sthaha) [reply pkg/components/modelserver/modelserver.go]: Yes, I am happy to have this validated later. 
Could you please add a todo .. something like  ` // TODO: ensure trailing , is accepted`

File: pkg/utils/k8s/k8s.go [hunk 28]
```diff
@@ -61,6 +61,16 @@ func (l StringMap) ToMap() map[string]string {
...
 }
 
+func (l StringMap) AddIfNotEmpty(k, v string) StringMap {
+	if k == "" {
+		return StringMap{}
+	}
+	if v != "" {
+		l[k] = v
+	}
+	return l
```
Comment (sthaha) [inline pkg/utils/k8s/k8s.go:71]: nit: please feel free to ignore
```suggestion
	if k != "" && v != "" {
		l[k] = v
	}
	return l
```

File: pkg/reconciler/runner.go [hunk 29]
```diff
@@ -33,6 +33,7 @@ type Runner struct {
 	Logger      logr.Logger
 }
 
+// TODO: make sure that model server container (deployment) is ready before creating kepler daemonset
```
Comment (sthaha) [inline pkg/reconciler/runner.go:36]: we don't need to do that ... we just have to indicate in the status that kepler `Available = false` until all resources are available.

File: pkg/controllers/kepler.go [hunk 30]
```diff
@@ -425,9 +434,27 @@ func exporterReconcilers(k *v1alpha1.Kepler, cluster k8s.Cluster) []reconciler.R
...
+	pvc := modelserver.NewPVC()
+
+	rs := updatersForResources(k,
+		// namespace scoped
+		cm,
+		deploy,
+		svc,
+		pvc,
+	)
+	return rs, nil
```
Comment (sthaha) [inline pkg/controllers/kepler.go:455]: We need to ensure that the `status` of `kepler` is properly updated based on the state of deployments.

File: tests/run-e2e.sh [hunk 31]
```diff
@@ -1,3 +1,4 @@
+# TODO: add integration test for all deployment choices (minimum, estimator-only, server-only, full)
```
Comment (sthaha) [inline tests/run-e2e.sh:1]: @sunya-ch IMHO as a requirement, All major features must have an e2e before merging. We can of course create follow up PRs for testing edge cases but we must have e2e tests to exercise the green and the most common red path at least.

File: pkg/api/v1alpha1/kepler_types.go [hunk 32]
```diff
@@ -78,18 +85,41 @@ type ModelServerSpec struct {
...
+}
+
+type EstimatorGroup struct {
+	Total      *EstimatorConfig `json:"total,omitempty"`
+	Components *EstimatorConfig `json:"components,omitempty"`
+}
+
+type EstimatorConfig struct {
+	// +kubebuilder:default=false
+	SidecarEnabled bool `json:"sidecar,omitempty"`
```
Comment (sthaha) [inline pkg/api/v1alpha1/kepler_types.go:114]: isn't `Enabled` this better? We don't need to be specific about `how` a feature is enabled. That is implementation detail. 

```suggestion
	Enabled bool `json:"enabled,omitempty"`
```

File: tests/run-e2e.sh (hunk 31 above)
Comment (sthaha) [reply tests/run-e2e.sh:1]: @sunya-ch , now that #237  has merged, you should find the  `oc` package quite useful

File: pkg/api/v1alpha1/kepler_types.go (hunk 32 above)
Comment (sunya-ch) [reply pkg/api/v1alpha1/kepler_types.go:114]: Kepler estimator could mean the local estimator and sidecar estimator. The config here (e.g., initURL) is a common for both estimator. If sidecar is not enabled, the local estimator inside kepler exporter will be used.

File: pkg/reconciler/runner.go (hunk 29 above)
Comment (sunya-ch) [reply pkg/reconciler/runner.go:36]: Could you provide me a specific guide on this or can I leave this work for the other contributor who can quickly add it? I will move todo line to another place such as line 351 in kepler.go.

File: pkg/api/v1alpha1/kepler_types.go (hunk 32 above)
Comment (sthaha) [reply pkg/api/v1alpha1/kepler_types.go:114]: How about we then specify that in the spec itself 
```
estimator: 
  enabled: true|false
  mode: sidecar|built-in 
```

Comment (sunya-ch) [reply pkg/api/v1alpha1/kepler_types.go:114]: We will have only with and without sidecar. I think it is okay to have only sidecar spec to be true or false.

**Comment by sunya-ch** [comment]
moved to https://github.com/sustainable-computing-io/kepler-operator/pull/322
